scalableminds/webknossos
# Fixed Volume Tracing while holding Shift

Holding shift while contour tracing volumes created weird artefacts. This was caused by a disconnect between the mouse handlers. Mouse down would start a contour but only under certain prerequisites. Move move, however, would always add points to the contour's point list. 
 
I added a new property `contourTracingMode` to sync the two. `contourTracingMode` has  tree states: `IDLE`, `ADD_TO_VOLUME` and `DELETE_FROM_VOLUME`.

I am not sure if this fixes the issue described here: https://discuss.webknossos.org/t/volume-tracing-deleted/775/11

### Mailable description of changes:
 - [TracingView] Fixed an issue when tracing volumes and holding the shift key. The tracing outline should no longer draw weird shapes

### Steps to test:
- Start volume tracing
- Both tracing by drawing a contour and brush mode should work correctly
- Hold shift key and try drawing, nothing should happen
- Shift + Left Click on an existing volume segment should switch the volume mode to annotating that id

### Issues:
- related to #2400 

------
